commit 4e2b80ff17784d6f547816bc13b76c58f52b2ab9
Author: Matt Merhar <mattmerhar@protonmail.com>
Date:   Sun Jun 28 21:40:03 2026 -0400

    airoha: fix HSUART baud calculations
    
    On the W1700K-2, where console is ttyS0, EFR32MG21 is ttyS1, and GPS is
    ttyS2, only console was working as expected.
    
    Before:
            Port  Requested Actual
            ttyS0 115200    115200
            ttyS1 230400    115200
            ttyS2 460800    921600
            ttyS2 230400    230400
    
    After:
            Port  Requested Actual
            ttyS0 115200    115200
            ttyS1 115200    115200
            ttyS2 115200    115200
            ttyS2 230400    230400
            ttyS2 460800    460800
            ttyS2 921600    921600
    
    Note that testing with the GPS shows only /1 and /2 work as expected
    for HSUART, even though /4 would in theory allow for the lower baud
    rates to be slightly more accurate.
    
    Additionally, refactor to use a proper set_divisor callback.
    
    Signed-off-by: Matt Merhar <mattmerhar@protonmail.com>

diff --git a/target/linux/airoha/dts/an7581.dtsi b/target/linux/airoha/dts/an7581.dtsi
index f9c1abea51..3317541e3c 100644
--- a/target/linux/airoha/dts/an7581.dtsi
+++ b/target/linux/airoha/dts/an7581.dtsi
@@ -408,18 +408,18 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 32 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
 
 		hsuart3: serial@1fbe1000 {
-			compatible = "airoha,en7523-uart";
+			compatible = "airoha,en7523-hsuart";
 			reg = <0x0 0x1fbe1000 0x0 0x40>;
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 54 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
@@ -438,7 +438,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
@@ -449,7 +449,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
diff --git a/target/linux/airoha/dts/an7583.dtsi b/target/linux/airoha/dts/an7583.dtsi
index 88a1e0a579..bb164c0b6b 100644
--- a/target/linux/airoha/dts/an7583.dtsi
+++ b/target/linux/airoha/dts/an7583.dtsi
@@ -494,7 +494,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 32 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
@@ -505,7 +505,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 54 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
@@ -516,7 +516,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
@@ -527,7 +527,7 @@
 			reg-io-width = <4>;
 			reg-shift = <2>;
 			interrupts = <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>;
-			clock-frequency = <7372800>;
+			clock-frequency = <20000000>;
 
 			status = "disabled";
 		};
diff --git a/target/linux/airoha/patches-6.18/886-uart-add-en7523-support.patch b/target/linux/airoha/patches-6.18/886-uart-add-en7523-support.patch
index 1118066013..1c881c229c 100644
--- a/target/linux/airoha/patches-6.18/886-uart-add-en7523-support.patch
+++ b/target/linux/airoha/patches-6.18/886-uart-add-en7523-support.patch
@@ -1,6 +1,6 @@
 --- /dev/null
 +++ b/drivers/tty/serial/8250/8250_en7523.c
-@@ -0,0 +1,96 @@
+@@ -0,0 +1,110 @@
 +// SPDX-License-Identifier: GPL-2.0+
 +/*
 + * Airoha EN7523 driver.
@@ -42,68 +42,95 @@
 + *
 + * Selecting divider needs to fulfill
 + * 1.8432 MHz <= xindiv_clk <= APB clock / 2
-+ * The clocks are unknown but a divider of value 1 did not work.
++ *
++ * The regular UART XINCLKDR register is an index into [1,2,4,8,10,12,16,20].
++ * The HSUART (airoha,en7523-hsuart) uses the XINCLKDR value as a direct
++ * clock divisor instead. Separate tables are used for each port type.
 + *
 + * Optimally the XYD, BRD and XINCLK_DIVCNT registers could be searched to
 + * find values that gives the least error for every baud rate. But searching
 + * the space takes time and in practise only a few rates are of interest.
 + * With some value combinations not working a tested subset is used giving
-+ * a usable range from 110 to 460800 baud.
++ * a usable range from 110 to 921600 baud.
 + */
 +
-+#define CLOCK_DIV_TAB_ELEMS 3
 +#define XYD_Y 65000
 +#define XINDIV_CLOCK 20000000
 +#define UART_BRDL_20M 0x01
 +#define UART_BRDH_20M 0x00
 +
-+static int clock_div_tab[] = { 10, 4, 2};
-+static int clock_div_reg[] = {  4, 2, 1};
++/* Regular UART: XINCLKDR is an index into [1,2,4,8,10,12,16,20] */
++static const int uart_div_tab[] = {10, 4, 2, 1};
++static const int uart_div_reg[] = {4, 2, 1, 0};
 +
++/* HSUART: XINCLKDR value is the clock divisor directly */
++static const int hsuart_div_tab[] = {2, 1};
++static const int hsuart_div_reg[] = {2, 1};
 +
-+int en7523_set_uart_baud_rate (struct uart_port *port, unsigned int baud)
++void en7523_set_divisor(struct uart_port *port, unsigned int baud,
++			unsigned int quot, unsigned int quot_frac)
 +{
 +	struct uart_8250_port *up = up_to_u8250p(port);
++	const int *div_tab, *div_reg;
 +	unsigned int xyd_x, nom, denom;
-+	int i;
++	int i, elems;
++
++	if (of_device_is_compatible(port->dev->of_node, "airoha,en7523-hsuart")) {
++		div_tab = hsuart_div_tab;
++		div_reg = hsuart_div_reg;
++		elems = ARRAY_SIZE(hsuart_div_tab);
++	} else {
++		div_tab = uart_div_tab;
++		div_reg = uart_div_reg;
++		elems = ARRAY_SIZE(uart_div_tab);
++	}
 +
 +	/* set DLAB to access the baud rate divider registers (BRDH, BRDL) */
 +	serial_port_out(port, UART_LCR, up->lcr | UART_LCR_DLAB);
 +
-+	/* set baud rate calculation defaults */
-+
 +	/* set BRDIV ([BRDH,BRDL]) to 1 */
 +	serial_port_out(port, UART_BRDL, UART_BRDL_20M);
 +	serial_port_out(port, UART_BRDH, UART_BRDH_20M);
 +
 +	/* calculate XYD_x and XINCLKDR register */
 +
-+	for (i = 0 ; i < CLOCK_DIV_TAB_ELEMS ; i++) {
-+		denom = (XINDIV_CLOCK/40) / clock_div_tab[i];
++	for (i = 0; i < elems; i++) {
++		denom = (XINDIV_CLOCK/40) / div_tab[i];
 +		nom = (baud * (XYD_Y/40));
 +		xyd_x = ((nom/denom) << 4);
 +		if (xyd_x < XYD_Y) break;
 +	}
-+	if (i >= CLOCK_DIV_TAB_ELEMS)
-+		i = CLOCK_DIV_TAB_ELEMS - 1;
++	if (i >= elems)
++		i = elems - 1;
 +
-+	serial_port_out(port, UART_XINCLKDR, clock_div_reg[i]);
++	serial_port_out(port, UART_XINCLKDR, div_reg[i]);
 +	serial_port_out(port, UART_XYD, (xyd_x<<16) | XYD_Y);
 +
 +	/* unset DLAB */
 +	serial_port_out(port, UART_LCR, up->lcr);
-+
-+	return 0;
 +}
 +
-+EXPORT_SYMBOL_GPL(en7523_set_uart_baud_rate);
++EXPORT_SYMBOL_GPL(en7523_set_divisor);
 --- a/drivers/tty/serial/8250/8250_of.c
 +++ b/drivers/tty/serial/8250/8250_of.c
-@@ -353,6 +353,7 @@ static const struct of_device_id of_plat
+@@ -173,6 +173,11 @@ static int of_platform_serial_setup(stru
+ 	case PORT_NPCM:
+ 		ret = npcm_setup(port);
+ 		break;
++#ifdef CONFIG_SERIAL_8250_AIROHA
++	case PORT_AIROHA:
++		port->set_divisor = en7523_set_divisor;
++		break;
++#endif
+ 	default:
+ 		/* Nothing to do */
+ 		ret = 0;
+@@ -353,6 +358,8 @@ static const struct of_device_id of_plat
  	{ .compatible = "ti,da830-uart", .data = (void *)PORT_DA830, },
  	{ .compatible = "nuvoton,wpcm450-uart", .data = (void *)PORT_NPCM, },
  	{ .compatible = "nuvoton,npcm750-uart", .data = (void *)PORT_NPCM, },
-+	{ .compatible = "airoha,en7523-uart", .data = (void *)PORT_AIROHA, },
++	{ .compatible = "airoha,en7523-uart",   .data = (void *)PORT_AIROHA, },
++	{ .compatible = "airoha,en7523-hsuart", .data = (void *)PORT_AIROHA, },
  	{ /* end of list */ },
  };
  MODULE_DEVICE_TABLE(of, of_platform_serial_table);
@@ -124,19 +151,6 @@
  };
  
  /* Uart divisor latch read */
-@@ -2769,6 +2777,12 @@ serial8250_do_set_termios(struct uart_po
- 	baud = serial8250_get_baud_rate(port, termios, old);
- 	quot = serial8250_get_divisor(port, baud, &frac);
- 
-+#ifdef CONFIG_SERIAL_8250_AIROHA
-+	/* Airoha SoCs have custom registers for baud rate settings */
-+	if (port->type == PORT_AIROHA)
-+		en7523_set_uart_baud_rate(port, baud);
-+#endif
-+
- 	/*
- 	 * Ok, we're now changing the port state. Do it with interrupts disabled.
- 	 *
 --- a/drivers/tty/serial/8250/Kconfig
 +++ b/drivers/tty/serial/8250/Kconfig
 @@ -356,6 +356,16 @@ config SERIAL_8250_ACORN
@@ -198,11 +212,12 @@
   * ARM specific type numbers.  These are not currently guaranteed
 --- a/include/linux/serial_8250.h
 +++ b/include/linux/serial_8250.h
-@@ -195,6 +195,7 @@ void serial8250_do_set_mctrl(struct uart
+@@ -195,6 +195,8 @@ void serial8250_do_set_mctrl(struct uart
  void serial8250_do_set_divisor(struct uart_port *port, unsigned int baud,
  			       unsigned int quot);
  int fsl8250_handle_irq(struct uart_port *port);
-+int en7523_set_uart_baud_rate(struct uart_port *port, unsigned int baud);
++void en7523_set_divisor(struct uart_port *port, unsigned int baud,
++			unsigned int quot, unsigned int quot_frac);
  void serial8250_handle_irq_locked(struct uart_port *port, unsigned int iir);
  int serial8250_handle_irq(struct uart_port *port, unsigned int iir);
  u16 serial8250_rx_chars(struct uart_8250_port *up, u16 lsr);
